File: rotten_oranges.cpp
// Time Complexity :O(mn)
// Space Complexity :O(mn)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : no


// Your code here along with comments explaining your approach
//BFS technique is used to fill up rotten oranges first and then we proceed level by level.
class Solution {
public:
    int orangesRotting(vector<vector<int>>& grid) {
        //edge case
        
        int row = grid.size();
        int col = grid[0].size();
        //BFS
        queue<vector<int>> my_index_queue;
        for(int i =0; i< row; i++){
            for(int j =0; j< col; j++){
                if(grid[i][j] ==2){
                 vector<int> temp = {i,j} ;
                 my_index_queue.push(temp); 
                }
            }
        }
        int size, time=0;
        int dirs[4][2]={{1,0},{-1,0},{0,-1},{0,1}};
        while(!my_index_queue.empty()){
            size = my_index_queue.size();
            while(size>0){
                vector<int> temp =  my_index_queue.front();
                my_index_queue.pop();
                size--;
                for(int i=0; i<4; i++){
                    int r = temp[0]+dirs[i][0];
                    int c = temp[1]+dirs[i][1];
                    if(r>=0 && r<row && c>=0 && c< col){
                       vector<int> temp2 = {r,c};
                       if(grid[r][c] == 1) {
                           my_index_queue.push(temp2); 
                           grid[r][c] = 2;
                       }
                    }
                    
                }
            }
            time++;
        }
        for(int i =0; i< row; i++){
            for(int j =0; j< col; j++){
                if(grid[i][j] ==1){
                 return -1;
                }
            }
        }
        if(time>0) return time-1;
        return time;
    }
};

File: rottenOranges.js
// Time Complexity : O(M*N)
// Space Complexity : O(M*N)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No



var orangesRotting = function(grid) {
    if(grid == null || grid.length == 0) return 0;
    let fresh = 0, m = grid.length, n = grid[0].length;

    let q = [];
    for(let  i = 0; i < m; i++){
        for(let j = 0; j < n; j++){
            if(grid[i][j] == 1) fresh++
            if(grid[i][j] == 1){
                q.push([i,j]);

            }
        }
    }

    let time = 0
    if(fresh == 0) return 0
    while(q.length){
        let size = q.length;
        for(let i = 0; i < size; i++){
            let [x, y] = q.shift();
            if (x - 1 >= 0 && grid[x - 1][y] === 1) {
                grid[x - 1][y] = 2;
                fresh--;
                queue.push([x - 1, y]);
              }
              if (x + 1 < height && grid[x + 1][y] === 1) {
                grid[x + 1][y] = 2;
                fresh--;
                queue.push([x + 1, y]);
              }
              if (y - 1 >= 0 && grid[x][y - 1] === 1) {
                grid[x][y - 1] = 2;
                fresh--;
                queue.push([x, y - 1]);
              }
              if (y + 1 < width && grid[x][y + 1] === 1) {
                grid[x][y + 1] = 2;
                fresh--;
                queue.push([x, y + 1]);
              }
        }
        if (queue.length > 0) minute++;
    }
    return fresh === 0 ? minute : -1;
};